src/mame/drivers/ksys573.cpp

```diff
@@ -710,6 +816,108 @@ void ksys573_state::gbbchmp_map(address_map& map)
 	map(0x1f640000, 0x1f640007).rw(m_duart, FUNC(mb89371_device::read), FUNC(mb89371_device::write)).umask32(0x00ff00ff);
 }
 
+bool ksys573_state::jvs_is_valid_packet()
+{
+    if (m_jvs_input_idx_w < 5) {
+		// A valid packet will have at the very least
+		//  - sync (0xe0)
+		//  - node number (non-zero)
+		//  - size
+		//  - at least 1 byte in the request message
+		//  - checksum
+		return false;
+    }
+
+	if (m_jvs_input_buffer[0] != 0xe0  || m_jvs_input_buffer[1] == 0x00) {
+		return false;
+	}
+
+    int command_size = m_jvs_input_buffer[2] + 3;
+    if (m_jvs_input_idx_w < command_size) {
+		return false;
+    }
+
+    uint8_t checksum = 0;
+    for (int i = 1; i < command_size - 1; i++) {
+		checksum += m_jvs_input_buffer[i];
+    }
+
+    return checksum == m_jvs_input_buffer[command_size - 1];
+}
+
+void ksys573_state::jvs_input_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+    m_jvs_input_buffer[m_jvs_input_idx_w++] = data & 0xff;
+    m_jvs_input_buffer[m_jvs_input_idx_w++] = data >> 8;
+
+	if (m_jvs_input_buffer[0] != 0xe0) {
+		m_jvs_input_idx_w = 0;
+	}
+
+    if (jvs_is_valid_packet()) {
+#ifdef JVS_DEBUG
+		printf("jvs_input_w( %08x, %08x, %02x %02x )\n", offset, mem_mask, data & 0xff, data >> 8 );
+
+		for (int i = 0; i < m_jvs_input_idx_w; i++) {
+			printf("%02x ", m_jvs_input_buffer[i]);
+		}
+
+		printf("\n");
+#endif
+
+		int command_size = m_jvs_input_buffer[2] + 3;
+		m_jvs_master->send_packet(m_jvs_input_buffer + 1, command_size - 2); // jvshost doesn't actually check the checksum, so don't send it
+
+		m_jvs_input_idx_w = 0;
+		m_jvs_input_idx_r = 0;
+
+		m_jvs_input_buffer[0] = 0;
+    }
+}
+
+uint16_t ksys573_state::jvs_input_r(offs_t offset, uint16_t mem_mask)
+{
+    uint16_t data = m_jvs_input_buffer[m_jvs_input_idx_r++];
+    data |= m_jvs_input_buffer[m_jvs_input_idx_r++] << 8;
+
+    return data;
+}
+
+uint16_t ksys573_state::port_in2_jvs_r(offs_t offset, uint16_t mem_mask)
+{
+	if (offset == 0) {
+		// 0x1f400008-0x1f400009 are for inputs
+		return ioport("IN2")->read();
```

Comment:
Please don’t add runtime tag map lookups.  They’re slow and you don’t benefit from validation.